pkg/tsdb/cloudwatch/live.go

```diff
@@ -0,0 +1,269 @@
+package cloudwatch
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/centrifugal/centrifuge"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/tsdb"
+	uuid "github.com/satori/go.uuid"
+	"golang.org/x/sync/errgroup"
+)
+
+type LogQueryRunnerSupplier struct {
+}
+
+type LogQueryRunner struct {
+	channelName string
+	publish     models.ChannelPublisher
+	running     map[string]bool
+}
+
+const (
+	maxAttempts         = 8
+	initialPollInterval = 500 * time.Millisecond
+)
+
+var (
+	channelMu        = sync.Mutex{}
+	responseChannels = map[string]chan *tsdb.Response{}
+)
+
+func addResponseChannel(name string, channel chan *tsdb.Response) error {
+	channelMu.Lock()
+	defer channelMu.Unlock()
+
+	if _, ok := responseChannels[name]; ok {
+		return fmt.Errorf("channel with name '%s' already exists", name)
+	}
+
+	responseChannels[name] = channel
+	return nil
+}
+
+func getResponseChannel(name string) (chan *tsdb.Response, error) {
+	channelMu.Lock()
+	defer channelMu.Unlock()
+
+	if responseChannel, ok := responseChannels[name]; ok {
+		return responseChannel, nil
+	}
+
+	return nil, fmt.Errorf("channel with name '%s' not found", name)
+}
+
+func deleteResponseChannel(name string) {
+	delete(responseChannels, name)
+}
```

Comment:
I would inline this function, it's too simple to warrant the abstraction.

Reply:
On reading the code more closely, inline it and avoid the global.

Reply:
Maybe I'm misunderstanding but why would inlining avoid using a global variable?